### ISSUE TITLE
getConnection method from PgConnection does not use configuration

### ISSUE THREAD

**Issue by rochoa** [CartoDB/Windshaft-cartodb#679]
Context: https://github.com/CartoDB/Windshaft-cartodb/blob/567928a7f50befe44e2bd063018e518eb5edf84c/lib/cartodb/backends/pg_connection.js#L111

The `getConnection` method should use the configuration to set, among other options, the max number of connections.

**Comment by oleurud** [comment]
Looking into the current configuration params (in addition to connection params), we have:

```
"extent":"xxxx",
"row_limit":xxxx,
"persist_connection":true/false,
"simplify_geometries":true/false,
"use_overviews":true/false,
"max_size":xxxx
```

and the PSQL could receive (in addition to connection params):
```javascript
 * @param {Object} poolParams
 * - size
 * - idleTimeout
 * - reapInterval
 * @param {Object} options
 * - {Number} maxRowSize The max number of bytes for a row, when exceeded the query will throw an error.
```

So, we can set the `size` (Max database connections in the pool). And it seems, no one more

**Comment by oleurud** [comment]
Branch created. We must do an intensive acceptance, similar to a load benchmark in order to ensure the right behavior due to the increment of the pool size

**Comment by rochoa** [comment]
That [`postgres`](https://github.com/CartoDB/Windshaft-cartodb/blob/42cd36afb7324ffae5f2b9b3711c101b8750be65/config/environments/development.js.example#L117-L144) attribute refers to Mapnik's PostGIS datasource. The Mapnik's PostGIS datasource use a different pool from the one we use —through node-postgres— for other PostgreSQL queries.

I would **not** use that pool size for node-postgres.

### The ideal solution

1. Create a new general configuration entry for PostgreSQL pool.
1. Move "`postgres`" attributes inside [Mapnik's renderer configuration](https://github.com/CartoDB/Windshaft-cartodb/blob/42cd36afb7324ffae5f2b9b3711c101b8750be65/config/environments/development.js.example#L172) to make it clear where that belongs.

### Unify PostgreSQL pool configurations

Apart from that, we could do some clean-up in the configuration. Right now, we define the pool configuration for different components, e.g. the [MVT renderer](https://github.com/CartoDB/Windshaft-cartodb/blob/42cd36afb7324ffae5f2b9b3711c101b8750be65/config/environments/development.js.example#L163-L171).

We should unify that, so in our production environment, it's possible to have the same configuration for all components.

#### Why?

The node-postgres version we use, `v6.1.x`, manages internally a multi-pool, [creating different pools for each different configuration object](https://github.com/brianc/node-postgres/blob/v6.1.6/lib/index.js#L57). That's very nice because it abstracts you from handling the pools, however, that means that if you throw a different combination of configuration values for the same user —let's say same database host and user, size set to 16, but different reapInterval— you will end with two different pools for a total of 2*16 connections, which is far from ideal.

That's why having just one source for the PostgreSQL pool configuration is so important. As important as having just one version of node-postgres in your environment, to avoid multiple pools for the same database/user.

Hope that helps.

**Comment by oleurud** [comment]
Thank you very much @rochoa And of course, your comment helps me a lot, increasing my knowledge of the current state and defining an objective

**Comment by rochoa** [comment]
VAMOS! 💪

**Comment by dgaubert** [comment]
I'm taking care of this issue since I'm in RT and @oleurud is working on `copyfrom/copyto` feature.

A couple of notes to take into account:

- In [Windhaft-cartodb](https://github.com/CartoDB/Windshaft-cartodb/search?utf8=%E2%9C%93&q=%22new+PSQL%22) we create `psql` clients always w/o passing `poolParams` argument. Indeed, we pass [conString](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L143) as connection config which is used to create the key for the pool object (note that it just uses `username`, `password`, `database`,  `dbhost` and  `dbport`). For the same user connection params we use the [default pool configuration](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L77-L85) through `pg.defaults` with [16 connections per pool](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L60). So the `poolSize` param never is used to compose the `poolName`. 

- However, in [Windshaft](https://github.com/CartoDB/Windshaft/search?utf8=✓&q="new+PSQL") we create `psql` clients with default pool params for attributes backend and custom pool params for the brand new postgis-mvt renderer. As postgis-mvt renderer is not being used in production yet, we can say that we are safe... for now. We are safe because the pool-params that we pass as the second argument in [node-cartodb-psql](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L55) overrides the [pg's defaults](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L79-L88) which is a singleton and after that, we will use a different pool-params configuration for incoming requests.

- On the other hand, as we can see in [`global.settings.environment.postgres`](https://github.com/CartoDB/Windshaft-cartodb/blob/master/config/environments/production.js.example#L138) there is a comment that claims that it's a config for mapnik datasource. But taking a [deeper look into it](https://github.com/CartoDB/Windshaft-cartodb/search?utf8=✓&q="environment.postgres"), we only use some properties (user, password, host and port) as default parameters to create a user or public user connections mainly.

- Regarding the pool parameters that we defined for different renderers (`mapnik`, `mvt`, `geojson` and `torque`). We use them for `mapnik`, `mvt` and `http` but we missed to pass the configuration for `torque` renderer. We didn't notice because it uses the internal `cartodb-psql` value for poolSize which is 16.

- Regarding the `poolSize` value for mapnik renderer is `8` for the production environment. This parameter could be misleading because it's going to be used by `tilelive-mapnik` as the number of maximum number of loaded maps at the very same time. Then, mapnik has an internal pool for Postgres connections with a maximum value of 10 (by default).

**Comment by rochoa** [comment]
> - In [Windhaft-cartodb](https://github.com/CartoDB/Windshaft-cartodb/search?utf8=%E2%9C%93&q=%22new+PSQL%22) we create `psql` clients always w/o passing `poolParams` argument. Indeed, we pass [conString](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L143) as connection config which is used to create the key for the pool object (note that it just uses `username`, `password`, `database`,  `dbhost` and  `dbport`). For the same user connection params we use the [default pool configuration](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L77-L85) through `pg.defaults` with [16 connections per pool](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L60). So the `poolSize` param never is used to compose the `poolName`. 

But [we could use the connection object](https://github.com/CartoDB/node-cartodb-psql/blob/139532a5704d1930b2744f37fd3a420f010186bf/lib/psql.js#L155).

> - However, in [Windshaft](https://github.com/CartoDB/Windshaft/search?utf8=✓&q="new+PSQL") we create `psql` clients with default pool params for attributes backend and custom pool params for the brand new postgis-mvt renderer. As postgis-mvt renderer is not being used in production yet, we can say that we are safe... for now. We are safe because the pool-params that we pass as the second argument in [node-cartodb-psql](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L55) overrides the [pg's defaults](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L79-L88) which is a singleton and after that, we will use a different pool-params configuration for incoming requests.

That was one my motivations to get this right before we go full throttle with that renderer.

> - On the other hand, as we can see in [`global.settings.environment.postgres`](https://github.com/CartoDB/Windshaft-cartodb/blob/master/config/environments/production.js.example#L138) there is a comment that claims that it's a config for mapnik datasource. But taking a [deeper look into it](https://github.com/CartoDB/Windshaft-cartodb/search?utf8=✓&q="environment.postgres"), we only use some properties (user, password, host and port) as default parameters to create a user or public user connections mainly.

We actually do [pass the `postgres` properties into grainstore datasource](https://github.com/CartoDB/Windshaft-cartodb/blob/dea1c74fcca12d0d80bada604d48253f9ceb4918/lib/cartodb/server_options.js#L121). Later, [we use those options to compose the XML for Mapnik](https://github.com/CartoDB/grainstore/blob/613e1e487d48dc1d64f8d5308ef101081ca0520d/lib/grainstore/mml-builder/mml-builder-inline.js#L103-L117).

> - Regarding the pool parameters that we defined for different renderers (`mapnik`, `mvt`, `geojson` and `torque`). We use them for `mapnik`, `mvt` and `http` but we missed to pass the configuration for `torque` renderer. We didn't notice because it uses the internal `cartodb-psql` value for poolSize which is 16.

So, this is in a similar to MVT Postgis renderer, isn't it?

> - Regarding the `poolSize` value for mapnik renderer is `8` for the production environment. This parameter could be misleading because it's going to be used by `tilelive-mapnik` as the number of maximum number of loaded maps at the very same time. Then, mapnik has an internal pool for Postgres connections with a maximum value of 10 (by default). 

As stated above, this is overriden by in grainstore with the `postgres` object.

HTH.

**Comment by dgaubert** [comment]
> We actually do pass the Postgres properties into grainstore datasource. Later, we use those options to compose the XML for Mapnik.

OMG, I totally missed that XML has the datasource configuration. Can you imagine the day that I'm able to keep every single detail of our stack in my mind?

Thanks for your comments @rochoa. Now, I have the whole picture to carry on with this issue.

**Comment by dgaubert** [comment]
For the record, `postgis` datasource parameters reference ->  https://github.com/mapnik/mapnik/wiki/PostGIS

**Comment by dgaubert** [comment]
Tasks to be achieved:
- Windsahft-cartodb:
  - [ ] Create a new general configuration entry for PostgreSQL pool.
  - [ ] Move "postgres" attributes inside Mapnik's renderer configuration to make it clear where that belongs.
  - [ ] Pass the very same pool configuration while creating windshaft's renderers
- Windshaft:
  - [ ] Unify configration poolParam we use to create PSQL client for different renderers.

**Comment by Algunenano** [comment]
Maybe it's a good time to have a look at [pg-pool](https://github.com/brianc/node-pg-pool) and how it integrates with node-postgres.

**Comment by dgaubert** [comment]
@Algunenano _don't throw the stone and hide your hand_ 😂 => What are you thinking about?

**Comment by Algunenano** [comment]
:moyai: :moyai: :moyai:

As I've been reviewing lately the changes in node-postgres I saw several references to the pools and a big change in v6:
```
v6.0.0
Breaking Changes
Remove pg.pools. There is still a reference kept to the pools created & tracked by pg.connect but it has been renamed, is considered private, and should not be used. Accessing this API directly was uncommon and was supposed to be private but was incorrectly documented on the wiki. Therefore, it is a breaking change of an (unintentionally) public interface to remove it by renaming it & making it private. Eventually pg.connect itself will be deprecated in favor of instantiating pools directly via new pg.Pool() so this property should become completely moot at some point. In the mean time...check out the new features...
New features
Replace internal pooling code with pg-pool. This is the first step in eventually deprecating and removing the singleton pg.connect. The pg-pool constructor is exported from node-postgres at require('pg').Pool. It provides a backwards compatible interface with pg.connect as well as a promise based interface & additional niceties.
You can now create an instance of a pool and don't have to rely on the pg singleton for anything:

var pg = require('pg')

var pool = new pg.Pool()

// your friendly neighborhood pool interface, without the singleton
pool.connect(function(err, client, done) {
  // ...
})
Promise support & other goodness lives now in pg-pool.

Please read the readme at pg-pool for the full api.
```

I haven't taken the time to check any changes but, as I think we should tend to be as close to upstream as possible, we could take the time and adapt our node-cartodb-psql module to use the latest (7) release to the fullest instead of relying in the old functions (`pg.connect`...).

**Comment by dgaubert** [comment]
Got it. Let me create a ticket with your proposal. ;)

**Comment by Algunenano** [comment]
Also somewhat related: I created a [PR](https://github.com/CartoDB/node-postgres/pull/5) to update the fork to 6.4.1 plus some improvements (pg10 support + test in travis, remove warnings) but if you feel it's superseded by a future update to node-postgres@7 we could drop it.

**Comment by dgaubert** [comment]
>> - In [Windhaft-cartodb](https://github.com/CartoDB/Windshaft-cartodb/search?utf8=%E2%9C%93&q=%22new+PSQL%22) we create `psql` clients always w/o passing `poolParams` argument. Indeed, we pass [conString](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L143) as connection config which is used to create the key for the pool object (note that it just uses `username`, `password`, `database`,  `dbhost` and  `dbport`). For the same user connection params we use the [default pool configuration](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L77-L85) through `pg.defaults` with [16 connections per pool](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L60). So the `poolSize` param never is used to compose the `poolName`. 

> But [we could use the connection object](https://github.com/CartoDB/node-cartodb-psql/blob/139532a5704d1930b2744f37fd3a420f010186bf/lib/psql.js#L155).

Hey @rochoa, 

Something that yesterday I realized is that actually `poolParams` is never used to create [the key for the pool object](https://github.com/CartoDB/node-postgres/blob/cdb-6.1/lib/index.js#L57) as it uses the configuration passed as argument that we built with [`that.getConnectionConfig()`](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L201). Later `node-postgres` overrides `poolSize` and `idleTimeoutMillis`. However, if the pool was created previously then the pool exists already and never gets the new pool configuration.

So basically we have one pool per combination of `user`, `password`, `database`, `host` and `port`. On the other hand, `poolSize` or `idleTimeoutMillis` will have the values that were used to create that pool for the first time and that pool.

My question is: do you think we should change this behaviour?. 

In my hummble opinion, we should change it and we should NOT modify [pg's defaults](https://github.com/CartoDB/node-cartodb-psql/blob/master/lib/psql.js#L79-L82) and *replace* existing pools if the pool configuration changes.
